src/mds/Server.cc

```diff
@@ -4175,6 +4169,12 @@ void Server::handle_client_getattr(const MDRequestRef& mdr, bool is_lookup)
       mdr->locking_state &= ~MutationImpl::ALL_LOCKED;
     }
   }
+  if ((mask & CEPH_CAP_AUTH_SHARED) && !(issued & CEPH_CAP_AUTH_EXCL))
+    lov.add_rdlock(&ref->authlock);
+  if ((mask & CEPH_CAP_LINK_SHARED) && !(issued & CEPH_CAP_LINK_EXCL))
+    lov.add_rdlock(&ref->linklock);
+  if ((mask & CEPH_CAP_XATTR_SHARED) && !(issued & CEPH_CAP_XATTR_EXCL))
+    lov.add_rdlock(&ref->xattrlock);
```

Comment:
We don't need to actually sort these when adding to the `LockOpVec`:

https://github.com/ceph/ceph/blob/97cb38d4da1472d533a7ccc00fa88192379f5191/src/mds/Mutation.cc#L116-L162

because the locks are sorted before acquisition.

I think what needs to happen for 

https://tracker.ceph.com/issues/62123

is to modify `Locker::acquire_locks` to look at the locks held by the mutation to ensure that it does not already hold locks (not necessarily in the `LockOpVec`) that would reverse the locking order if the locks in the `LockOpVec` are acquired.


That is probably an expensive consistency check so it would need to be a debug config that turns it on.

Reply:
Yeah, you are correct. I missed this part. I will have a look about this again. Thanks @batrick

Reply:
> That is probably an expensive consistency check so it would need to be a debug config that turns it on.

Not just that, the check should be configurable with a flag on the acquire locks call to be able to suppress it for some edge cases where we know what we are doing.

Reply:
Actually the issue here is that when a single request triggers two 'acquire_locks()' then it will be possibly break the lock-order and then cause lock-order deadlocks between multiple requests.
Anyway we should start over the request when the second 'acquire_locks()' fails as @batrick did in https://github.com/ceph/ceph/pull/52522, which I think @batrick just fixed one part of it, not all.

@batrick @leonid-s-usov I have pushed a new commit to fix the `getattr` and `setattr` only for now, please check and if this is the correct direction I will continue to fix all the other cases.

Thanks.